File: cpp/src/ev1/ut2_controlflujometodos/u03bucles/teoria/U02_While.cpp
/*
U02 — Bucle while: repetir mientras la condicion sea verdadera.

OBJETIVO:
  - Usar while cuando no sabemos cuantas vueltas haran falta.
  - Implementar un contador manual dentro del while.
  - Conocer el patron centinela (valor especial que detiene el bucle).

Autor: Agustin. A. Marquez. Pina
Contacto: agu1406@outlook.es
Repositorio GitHub: https://github.com/Agu1406/ClasesParticulares
Sitio web: https://www.agustinmarquez.dev
*/

#include <iostream>
using namespace std;

int main() {
    cout << "\n¡INICIO DEL PROGRAMA!\n" << endl;

    /*
    PRIMERA PARTE - while con contador (equivalente a un for simple).
      Importante: modificar la variable de control dentro del bucle (n++).
    */
    cout << "¡DEMO 1: CONTADOR CON WHILE!\n" << endl;

    int n = 1;
    while (n <= 5)
    {
        cout << "  n = " << n << endl;
        n++;
    }

    /*
    SEGUNDA PARTE - Acumular hasta llegar a un objetivo.
      Se repite mientras la suma no supere 100.
    */
    cout << "\n¡DEMO 2: SUMAR HASTA SUPERAR 100!\n" << endl;

    int suma = 0;
    int termino = 1;
    while (suma <= 100)
    {
        suma += termino;
        cout << "  +" << termino << " -> suma parcial = " << suma << endl;
        termino++;
    }
    cout << "Se detuvo con suma = " << suma << "\n" << endl;

    /*
    TERCERA PARTE - Patron centinela con datos simulados.
      Un centinela es un valor especial (-1) que marca el final de datos.
      (En ejercicios reales vendria de cin).
    */
    cout << "¡DEMO 3: CENTINELA -1!\n" << endl;

    int dato1 = 12;
    int dato2 = 45;
    int dato3 = 8;
    int dato4 = 33;
    int dato5 = -1;  // centinela: fin de datos
    int paso = 1;
    int valor = dato1;

    cout << "Valores leidos: ";
    while (valor != -1)  // centinela: cuando es -1, paramos
    {
        cout << valor << " ";

        paso++;
        if (paso == 2)
        {
            valor = dato2;
        }
        else if (paso == 3)
        {
            valor = dato3;
        }
        else if (paso == 4)
        {
            valor = dato4;
        }
        else
        {
            valor = dato5;
        }
    }
    cout << "(centinela -1 detectado, bucle terminado)" << endl;

    /*
    CUARTA PARTE - Buscar el primer multiplo de 7 mayor que 50.
    */
    cout << "\n¡DEMO 4: BUSQUEDA CON WHILE!\n" << endl;

    int candidato = 1;
    while (candidato <= 50 || candidato % 7 != 0)
    {
        candidato++;
    }
    cout << "Primer multiplo de 7 despues de 50: " << candidato << endl;

    cout << "\n¡FIN DEL PROGRAMA!\n" << endl;
    return 0;
}

File: python/src/ev1/ut2_controlflujometodos/u02bucles/teoria/U02_While.py
"""
U02 — Bucle while.

OBJETIVO:
  - Repetir mientras una condicion sea True.
  - Actualizar la variable de control dentro del bucle (evitar infinito).
  - break sale del bucle; continue salta a la siguiente vuelta.

Autor: Agustin. A. Marquez. Pina
Contacto: agu1406@outlook.es
Repositorio GitHub: https://github.com/Agu1406/ClasesParticulares
Sitio web: https://www.agustinmarquez.dev
"""
print("\n¡INICIO DEL PROGRAMA!\n")
"""
PRIMERA PARTE - Sintaxis y uso del bucle "while".
"""
contador = 0
while contador < 5:
    # print(): muestra el valor actual del contador
    print("contador =", contador)
    contador += 1

suma = 0
numero = 1
while suma < 20:
    suma += numero
    # print(): f-string con el numero sumado y el total parcial
    print(f"sumando {numero}, total = {suma}")
    numero += 1

# print(): titulo del ejemplo con break
print("\nBuscando el primer multiplo de 7 mayor que 20:")
n = 21
while True:
    if n % 7 == 0:
        # print(): encontro el multiplo
        print("Encontrado:", n)
        # break: sale del while True
        break
    n += 1

# print(): titulo del ejemplo con continue
print("\nNumeros del 1 al 8 excepto el 5:")
i = 0
while i < 8:
    i += 1
    if i == 5:
        # continue: salta el print de abajo y vuelve al while
        continue
    # print(): muestra i en la misma linea
    print(i, end=" ")
# print(): salto de linea final
print()

secreto = 7
# for intento in range(1, 11): prueba numeros del 1 al 10 sin usar listas.
for intento in range(1, 11):
    # print(): muestra el intento actual
    print("Intento:", intento)
    if intento == secreto:
        # print(): acerto
        print("Correcto!")
        # break: deja el bucle for
        break
    # print(): no acerto esta vuelta
    print("Fallo, sigue...")
else:
    # else del for: solo si no hubo break
    print("No acertaste en 10 intentos")

print("\n¡FIN DEL PROGRAMA!\n")
